daemon/gluster-blockd.c

```diff
@@ -42,6 +44,69 @@ static gbConfig *gbCfg;
 
 struct timeval TIMEOUT = {CLI_TIMEOUT_DEF, 0};  /* remote rpc call timeout, 5 mins */
 
+typedef struct globalData {
+  int    pfd;  /* pidfile file descriptor */
+  pid_t  chpid;
+} globalData;
+
+globalData ctx = {0,};
+
+
+void
+glusterBlockCleanGlobals(void)
+{
+  struct flock lock = {0, };
+
+    lock.l_type = F_UNLCK;
+    if (fcntl(ctx.pfd, F_SETLK, &lock) == -1) {
+      LOG("mgmt", GB_LOG_ERROR, "fcntl(UNLCK) on pidfile %s failed[%s]",
+          GB_LOCK_FILE, strerror(errno));
+    }
+
+    if (ctx.pfd) {
+      close(ctx.pfd);
+    }
+    glusterBlockDestroyConfig(gbCfg);
+}
+
+
+void onSigServerHandler(int signum)
+{
+  LOG("mgmt", GB_LOG_DEBUG,
+      "server process received signal:%d with pid:%lu", signum, getpid());
+
+  exit(EXIT_SUCCESS);  /* server process terminated by signal */
+}
+
+
+void onSigCliHandler(int signum)
+{
+  int wstatus;
+
+
+  LOG("mgmt", GB_LOG_DEBUG,
+      "cli process received signal:%d with pid:%lu", signum, getpid());
+
+  kill(ctx.chpid, signum);  /* Pass the signal to server process */
+  waitpid(ctx.chpid, &wstatus, 0);
+
```

Comment:
And also here use the svc_stop, then just return and the cli process will also do the clean up in libtirpc.
Then the cli process will help waitpid for the child process in LINE 602.